#QTM 350 HW2: pandas Package
Ash Addanki, Jessica Kang, Dillon Wu, Tony Yang

![alt text](https://drive.google.com/uc?id=1BaBgT5w1dAvFd9ENaqrMqjgPa7so8TwD)


pandas is a Python package for data manipulation and analysis. Data can be stored in 1-dimensional arrays called series or in 2-dimensional Dataframes for tabular data. It is similar to R's Tidyverse in function. Knowing how to use pandas is important to being a more well-rounded data scientist as it is applicable for all fields and concentrations. For our group, we have a diverse set of fields in economics, history, computer science, and english. 

In [0]:
#Installing Pandas
pip install Pandas

In [0]:
#import pandas package
import pandas as pd


###History
Text analysis, or text mining, can be used in history to understand and sort long sequences of text or multiple works in cases where humans 
can't read such a large amount of information. Users can derive general trends of historical documents or the mood and tone of a certain author using text analysis. 
To show how pandas can be used in the process of analyzing historical documents, the Gettysburg Address will be used an example. 

The str.len() function counts the number of characters in a Series, which can be used to count the number of words in historical documents. 


In [0]:
#load Gettysburg Address and split it into its 3 paragraphs
address = pd.Series(['Four score and seven years ago our fathers brought forth on this continent, a new nation, conceived in Liberty, and dedicated to the proposition that all men are created equal.','Now we are engaged in a great civil war, testing whether that nation, or any nation so conceived and so dedicated, can long endure. We are met on a great battle-field of that war. We have come to dedicate a portion of that field, as a final resting place for those who here gave their lives that that nation might live. It is altogether fitting and proper that we should do this.','But, in a larger sense, we can not dedicate—we can not consecrate—we can not hallow—this ground. The brave men, living and dead, who struggled here, have consecrated it, far above our poor power to add or detract. The world will little note, nor long remember what we say here, but it can never forget what they did here. It is for us the living, rather, to be dedicated here to the unfinished work which they who fought here have thus far so nobly advanced. It is rather for us to be here dedicated to the great task remaining before us—that from these honored dead we take increased devotion to that cause for which they gave the last full measure of devotion—that we here highly resolve that these dead shall not have died in vain—that this nation, under God, shall have a new birth of freedom—and that government of the people, by the people, for the people, shall not perish from the earth.' ])

#turn all letters into lowercase to make all words the same case for comparison
address_lower = address.str.lower()

#compare lengths of each paragraph based on how many words there are
address.str.len()

0    176
1    379
2    895
dtype: int64

Let's say you want to find all instances where Lincoln directly talks about or references slavery. To compare, we can also find all instances where he mentions computers and instances of the Civil War. This lets us figure out what the document is about without having to read it. However, since the Gettysburg Addres is a very short speech, reading it personally would be more beneficial than using text analysis. 

In [0]:
slavery_words = ["slave", "slaves", "slavery"]
war_words = ["war", "civil", "battle", "dead", "fight"]
computer_words = ["computer", "laptop", "phone"]

slavery_count = address.str.count("slave") + address.str.count("slaves") + address.str.count("slavery")
war_count = address.str.count("war") + address.str.count("civil") + address.str.count("battle") + address.str.count("dead") + address.str.count("fight") 
computer_count = address.str.count("computer") + address.str.count("laptop") + address.str.count("phone")

print(slavery_count)
print(war_count)
print(computer_count)


0    0
1    0
2    0
dtype: int64
0    0
1    4
2    3
dtype: int64
0    0
1    0
2    0
dtype: int64


The Gettysburg returns 0 words in all 3 paragraphs about slavery and computers while it returns 7 words total with words related to war. This suggests that Lincoln's speech was neither about slavery nor computers, but was about war instead. This kind of analysis is very dangerous without the relevant background knowledge on historical events because the civil war was very much about slavery. However, knowing the context of the document is even more important to not misinterpet the results. There are more complicated methods for text analysis, such as sentiment analysis, or tf-idf, but the pandas package is not suited for that. 

# English


The pandas package is known to be great for text analysis and manipulation. One important format of data in Pandas is the Series, which is essentially a column of data. The data can be integers, objects, string, or anything. There are some particularly fundamental functions that come in handy when working with text. Here, we can explore some. 

In [0]:
# import pandas 
import pandas as pd 

The series.get() function retrieves the object of a given index value. This can be useful in many different ways. A way it can be used in text analysis is through identifying synonyms. 

In [0]:
# construct an index and series for demonstration 
original= pd.Series(['understand', 'listen', 'good','reason', 'pretty'])
synonym= pd.Series(['comprehend','hear', 'fine','cause', 'beautiful'])

In [0]:
# identify the index 
synonym.index=original

In [0]:
# view the series 
print(synonym)

understand    comprehend
listen              hear
good                fine
reason             cause
pretty         beautiful
dtype: object


In [0]:
# use the series.get function to call the object for a given index value 
result=synonym.get(key = 'reason')
print(result)

cause


The series.where() function replaces the object of a given index value if a condition needs to be met. If we continue with the series of synonyms, we can see more manipulation of text. 

In [0]:
# construct another index and series for demonstration 
original_2= pd.Series(['understand', 'listen', 'good','reason', 'pretty'])
synonym_2= pd.Series(['grasp','hear', 'fine','cause', 'lovely'])

In [0]:
# identify the index 
synonym_2.index=original_2

In [0]:
# view the series 
print(synonym_2)

understand     grasp
listen          hear
good            fine
reason         cause
pretty        lovely
dtype: object


In [0]:
# use the series.where function to replace the values
synonym.where(synonym == 'lovely', synonym_2)

understand     grasp
listen          hear
good            fine
reason         cause
pretty        lovely
dtype: object

## Economics

Pandas is great for working with large datasets. We can see how pandas can be used with an example dataset from Kaggle of video game sales. Let's say we're interested in comparing total sales revenue in North America and Europe for games released in 2015 to the sales revenue from games released in the earliest year in the dataset.  

See https://github.com/jkang4/QTM350-pandas

In [0]:
# import pandas
import pandas as pd

In [0]:
# read and assign the dataset as vidgames
vidgames = pd.read_csv('https://raw.githubusercontent.com/jkang4/QTM350-pandas/master/vgsales.csv')

In [0]:
vidgames.head()

Rank                      Name Platform  ...  JP_Sales Other_Sales Global_Sales
0     1                Wii Sports      Wii  ...      3.77        8.46        82.74
1     2         Super Mario Bros.      NES  ...      6.81        0.77        40.24
2     3            Mario Kart Wii      Wii  ...      3.79        3.31        35.82
3     4         Wii Sports Resort      Wii  ...      3.28        2.96        33.00
4     5  Pokemon Red/Pokemon Blue       GB  ...     10.22        1.00        31.37

[5 rows x 11 columns]

We can use `df.describe()` to see summary statistics for all the numerical columns so we have an idea of the distribution of each variable before we even start. We also use `df.min()` to find the earliest year that is included. 

In [0]:
vidgames.describe()

Rank          Year  ...   Other_Sales  Global_Sales
count  16598.000000  16327.000000  ...  16598.000000  16598.000000
mean    8300.605254   2006.406443  ...      0.048063      0.537441
std     4791.853933      5.828981  ...      0.188588      1.555028
min        1.000000   1980.000000  ...      0.000000      0.010000
25%     4151.250000   2003.000000  ...      0.000000      0.060000
50%     8300.500000   2007.000000  ...      0.010000      0.170000
75%    12449.750000   2010.000000  ...      0.040000      0.470000
max    16600.000000   2020.000000  ...     10.570000     82.740000

[8 rows x 7 columns]

In [0]:
vidgames["Year"].min()

1980.0

In this dataset, we have separate columns for sales in North America and Europe, so we want to make a new column that sums NA_Sales and EU_Sales. We specify columns with `df["col_name"]` so we select those two and name our new column.

In [0]:
vidgames["NA_EU_Sales"] = vidgames["NA_Sales"] + vidgames["EU_Sales"]
vidgames.head()

Rank                      Name  ... Global_Sales  NA_EU_Sales
0     1                Wii Sports  ...        82.74        70.51
1     2         Super Mario Bros.  ...        40.24        32.66
2     3            Mario Kart Wii  ...        35.82        28.73
3     4         Wii Sports Resort  ...        33.00        26.76
4     5  Pokemon Red/Pokemon Blue  ...        31.37        20.16

[5 rows x 12 columns]

Now let's only focus on the games from 1980 and 2015 by first creating a new dataframe with only the Year and NA_EU_Sales columns and then only including rows from those two years. 

In [0]:
# Create new dataframe with Year and NA_EA_Sales
vg2 = vidgames.filter(items=['Year', 'NA_EU_Sales'])
vg2.head()


Year  NA_EU_Sales
0  2006.0        70.51
1  1985.0        32.66
2  2008.0        28.73
3  2009.0        26.76
4  1996.0        20.16

In [0]:
# Select rows from 1980 or 2015
vidgamesrev = vg2.loc[(vg2['Year'] == 1980) | (vg2['Year'] == 2015)]
vidgamesrev

Year  NA_EU_Sales
33     2015.0        11.58
77     2015.0         7.17
92     2015.0         6.22
101    2015.0         6.61
109    2015.0         5.62
...       ...          ...
16499  2015.0         0.00
16528  2015.0         0.00
16542  2015.0         0.00
16548  2015.0         0.00
16565  2015.0         0.01

[623 rows x 2 columns]

We add all of the sales for each year.  

In [0]:
vidgamesrev.loc[vidgamesrev['Year'] == 1980].sum(axis=0)


Year           17820.00
NA_EU_Sales       11.26
dtype: float64

In [0]:
vidgamesrev.loc[vidgamesrev['Year'] == 2015].sum(axis=0)

Year           1237210.00
NA_EU_Sales        200.53
dtype: float64

Finally, we see that the total North American and European sales for 1980 video games is 11.26 million, while 2015 video games brought in almost 20 times that amount with 200.53 million.

## Data Cleaning

The titanic dataset is a dataset that can be downloaded here: https://www.kaggle.com/c/titanic. It is an introductory dataset used for machine learning instruction. In this section, we conduct data cleaning on the dataset.

In [0]:
# You will need to download the dataset from the website.
from google.colab import files
upload1 = files.upload()
upload2 = files.upload()

Saving train.csv to train (2).csv


Saving test.csv to test (2).csv


In [0]:
import io
train_orig = pd.read_csv(io.BytesIO(upload1['train.csv']))
test = pd.read_csv(io.BytesIO(upload2['test.csv']))
train = train_orig.copy()
# Looking at which columns have null values
train.isnull().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [0]:
# Looking at columns with null values in test dataset
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Here, we drop the column Cabin since there are too many null values to consider.

In [0]:
# Dropping Cabin column
train.drop(['Cabin'], axis = 1, inplace=True)

We show how one hot encoding can be done with pandas below. One hot encoding is a powerful tool used to transform a column of categorical values with many labels into columns where each column is a binary value. For example, in the transformation of the Ticket column in the Titanic dataset, we get rid of the Ticket column and add columns where each column title was originally a ticket label in the column. A row contains a 1 if that row originally had that ticket label, and 0 if otherwis. This transformation is useful for training machine learning models that have a difficult time handling categorical data. 

In [0]:
# One hot encoding 
train_temp=train.copy()
dummy = pd.get_dummies(train['Ticket'])
train_temp = pd.concat([train, dummy], axis=1, sort=False)
train_temp.drop(['Ticket'],axis=1,inplace=True)
train_temp

PassengerId  Survived  Pclass  ... W.E.P. 5734 W/C 14208  WE/P 5735
0              1         0       3  ...           0         0          0
1              2         1       1  ...           0         0          0
2              3         1       3  ...           0         0          0
3              4         1       1  ...           0         0          0
4              5         0       3  ...           0         0          0
..           ...       ...     ...  ...         ...       ...        ...
886          887         0       2  ...           0         0          0
887          888         1       1  ...           0         0          0
888          889         0       3  ...           0         0          0
889          890         1       1  ...           0         0          0
890          891         0       3  ...           0         0          0

[891 rows x 691 columns]